dracut/driver_updates.py

```diff
@@ -320,11 +320,46 @@ def grab_driver_files(outdir="/updates"):
     move_files(firmware, outdir+FIRMWARE_UPDATES_DIR, outdir+'/lib/firmware')
     return [os.path.basename(m).split('.ko')[0] for m in modules]
 
+def get_all_loaded_modules():
+    """parse /proc/modules for all loaded kernel modules"""
+    all_modules = []
+    with open("/proc/modules", "r") as modules:
+        line = modules.readline()
+        while line:
+            module_name = line.split(" ")[0]
+            all_modules.append(module_name)
+            line = modules.readline()
+    return all_modules
+
 def load_drivers(modnames):
-    """run depmod and try to modprobe all the given module names."""
+    """load all drivers based on given aliases. In case the drivers are
+    already present in the kernel, replace them with the new ones.
+    """
     log.debug("load_drivers: %s", modnames)
+    # save snapshot of currently installed modules
+    all_modules_org = get_all_loaded_modules()
+    # find and unload already loaded modules
+    # (Replacement modules are not known, until 'depmod -a' is called)
+    modules_to_replace = [x for x in modnames if x in all_modules_org]
+    # unload all marked modules as well as dependant modules
+    subprocess.call(["modprobe", "-r"] + modules_to_replace)
+    if modules_to_replace == []:
+        log.debug("replace_drivers: %s", modules_to_replace)
+
+    # original modules are all unloaded now
+    # refresh modules tree (i.e. switch to new modules)
     subprocess.call(["depmod", "-a"])
-    subprocess.call(["modprobe", "-a"] + modnames)
+    # add new modules
+    subprocess.call(["modprobe", "-a"] + modules_to_replace)
+
+    # get new snapshot of currently installed modules
+    all_modules_new = get_all_loaded_modules()
+    # compare snapshots and get modules removed from system due to dependencies
+    diff_modules = list(set(all_modules_org) - set(all_modules_new))
+    modules_to_add = list(set(diff_modules).union(set(modnames)))
+
+    # load all previously removed modules again
+    subprocess.call(["modprobe", "-a"] + modules_to_add)
```

Comment:
modules_to_add is merged (using union) with modnames which is mandatory non-empty parameter of the function. This guarantees list nonemptiness.